prow/cmd/phaino/local.go

```diff
@@ -0,0 +1,339 @@
+}
+
+var baseArgs = []string{"docker", "run", "--rm=true"}
+
+func checkPrivilege(cont coreapi.Container, allow bool) (bool, error) {
+	if cont.SecurityContext == nil {
+		return false, nil
+	}
+	if cont.SecurityContext.Privileged == nil {
+		return false, nil
+	}
+	if !*cont.SecurityContext.Privileged {
+		return false, nil
+	}
+	fmt.Fprint(os.Stderr, "Privileged jobs are unsafe. Remove from local run? [yes]: ")
+	var out string
+	fmt.Scanln(&out)
+	if out == "no" || out == "n" {
+		if !allow {
+			return false, errors.New("privileged jobs are disallowed")
+		}
+		logrus.Warn("DANGER: privileged containers are unsafe security risks. Please refactor")
+		return true, nil
+	}
+	return false, nil
+}
+
+func convertToLocal(log *logrus.Entry, pj prowapi.ProwJob, allowPrivilege bool) ([]string, error) {
+	log.Info("Converting job into docker run command...")
+	var localArgs []string
+	localArgs = append(localArgs, baseArgs...)
+	container := pj.Spec.PodSpec.Containers[0]
+	decoration := pj.Spec.DecorationConfig
+	var entrypoint string
+	args := container.Command
+	args = append(args, container.Args...)
+	if len(args) > 0 && decoration != nil {
+		entrypoint = args[0]
+		args = args[1:]
+	}
+	if entrypoint == "" && decoration != nil {
+		return nil, errors.New("decorated jobs must specify command and/or args")
+	}
+	if entrypoint != "" {
+		localArgs = append(localArgs, "--entrypoint="+entrypoint)
+	}
+
+	for _, env := range container.Env {
+		localArgs = append(localArgs, "-e", env.Name+"="+env.Value)
+	}
+
+	priv, err := checkPrivilege(container, allowPrivilege)
+	if err != nil {
+		return nil, err
+	}
+	if priv {
+		localArgs = append(localArgs, "--privileged")
+	}
+
+	if container.Resources.Requests != nil {
+		// TODO(fejta): https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources
+		log.Warn("Ignoring resource requirements")
+	}
+
+	for _, mount := range container.VolumeMounts {
+		vol := volume(*pj.Spec.PodSpec, mount.Name)
+		if vol == nil {
+			return nil, fmt.Errorf("mount %q missing associated volume", mount.Name)
+		}
+		if vol.EmptyDir != nil {
+			localArgs = append(localArgs, "-v", mount.MountPath)
+		} else {
+			local, err := readMount(mount)
+			if err != nil {
+				return nil, fmt.Errorf("bad mount %q: %v", mount.Name, err)
+			}
+			arg := local + ":" + mount.MountPath
+			if mount.ReadOnly {
+				arg += ":ro"
+			}
+			localArgs = append(localArgs, "-v", arg)
+		}
+	}
+
+	var workingDir string
+
+	if decoration != nil {
+		var refs []prowapi.Refs
+		if pj.Spec.Refs != nil {
+			refs = append(refs, *pj.Spec.Refs)
+		}
+		refs = append(refs, pj.Spec.ExtraRefs...)
+		for _, ref := range refs {
+			path := pathAlias(ref)
+			repo, err := readRepo(path)
+			if err != nil {
+				return nil, fmt.Errorf("bad %q repo: %v", path, err)
+			}
+			dest := filepath.Join("/go/src", path)
+			if workingDir == "" {
+				workingDir = dest
+			}
+			localArgs = append(localArgs, "-v", repo+":"+dest)
```

Comment:
This ignores `BaseRef`, `BaseSHA`, and the `Pulls` fields of the `Refs`.  Ideally we would respect these, but at a minimum we should warn that they are not respected if they are specified.

Reply:
Don't think this is necessary as we explicitly ask the user to provide the source for this ref.

Agree that long-term we should support a mode where we check out the repos for you (probably after we add knative-build job support, which will also have to deal with multiple containers).